src/main/java/com/minecolonies/coremod/commands/ColonyTPCommand.java

```diff
@@ -0,0 +1,236 @@
+
+            boolean foundPosition = isPositionSave(sender, groundPosition) && !findColony(groundPosition, sender.getEntityWorld());
+
+            /* Take the return and determine if good or bad */
+             /* send info to look to see if another colony is near */
+
+            if (foundPosition)
+            {
+                /* everything checks out good make the TP and jump out*/
+                playerToTeleport.setPositionAndUpdate(groundPosition.getX(), groundPosition.getY() + SAFETY_DROP, groundPosition.getZ());
+                return;
+            }
+        }
+        playerToTeleport.getCommandSenderEntity().addChatMessage(new TextComponentString("Couldn't find a safe spot.  Try again in a moment."));
+    }
+
+    /**
+     * this checks that you are not in the air or underground
+     * and if so it will look up and down for a good landing spot
+     * before TP
+     *
+     * @param blockPos for the current block LOC
+     * @return blockPos to be used for the TP
+     */
+    private static BlockPos findLand(final BlockPos blockPos, final World world)
+    {
+        int top = STARTING_Y;
+        int bot = 0;
+        int mid = STARTING_Y;
+
+        BlockPos foundland = null;
+        BlockPos tempPos = blockPos;
+        //We are doing a binary search to limit the amount of checks (usually at most 9 this way)
+        while (top >= bot)
+        {
+            tempPos = new BlockPos( tempPos.getX(),mid, tempPos.getZ());
+            Block blocks = world.getBlockState(tempPos).getBlock();
+            if (blocks == Blocks.AIR && world.canSeeSky(tempPos))
+            {
+                top = mid - 1;
+                foundland = tempPos;
+            }
+            else
+            {
+                bot = mid + 1;
+                foundland = tempPos;
+            }
+            mid = (bot + top)/2;
+        }
+
+        return foundland;
+    }
+
+    /**
+     * this checks that you are not in liquid.  Will check for all liquids, even those from other mods
+     * before TP
+     *
+     * @param blockPos for the current block LOC
+     * @param sender uses the player to get the world
+     * @return isSafe true=safe false=water or lava
+     */
+    private static boolean isPositionSave(@NotNull ICommandSender sender, BlockPos blockPos)
+    {
+        return sender.getEntityWorld().getBlockState(blockPos).getBlock() != Blocks.AIR && !sender.getEntityWorld().getBlockState(blockPos).getMaterial().isLiquid();
+    }
+    /**
+     * this checks that you are not too close to another colony
+     * before TP
+     *
+     * @param blockPos for the current block LOC
+     * @return colNear false=no true=yes
+     */
+    private boolean findColony(BlockPos blockPos, World world)
+    {
+        Colony nearestCol = ColonyManager.getClosestColony(world, blockPos);
+        Boolean colNear;
+        /* get individual coords to do the math */
+        int cx = nearestCol != null ? nearestCol.getCenter().getX() : 0;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Avoid if (x != y) ..; else ..; [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 'cx' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)